torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -1299,9 +1324,97 @@ void FuseTensorExprs(
+  //       custom_lowerings;
+  //   kernel = std::make_shared<tensorexpr::TensorExprKernel>(
+  //       node->g(attr::Subgraph), custom_lowerings, sym_shapes);
+  //   cached_kernels[node] = kernel;
+  // }
+  // return [kernel](Stack& stack) {
+  //   RECORD_FUNCTION(kernel->getKernelName(), std::vector<c10::IValue>());
+  //   kernel->run(stack);
+  //   return 0;
+  // };
+
+  // The following results in 2 different prim::TensorExprGroup nodes having
+  // the same address. Hence results in an error.
+  // return [=](Stack& stack) {
+  //   std::cout << "-- In createTensorExprOp: " << node << " : " << *node << std::endl;
+  //   static std::unordered_map<const Node*, std::shared_ptr<tensorexpr::TensorExprKernel>> cached_kernels;
+  //   auto it = cached_kernels.find(node);
+  //   std::shared_ptr<tensorexpr::TensorExprKernel> kernel;
+  //   if (it != cached_kernels.end()) {
+  //     kernel = it->second;
+  //     std::cout << "---- Reusing existing kernel" << std::endl;
+  //   } else {
+  //     std::cout << "---- Compiling a new kernel" << std::endl;
+  //     auto sym_shapes = node->is(attr::symbolic_shape_inputs);
+  //     std::unordered_map<c10::Symbol, tensorexpr::NNCLoweringFunction>
+  //         custom_lowerings;
+  //     kernel = std::make_shared<tensorexpr::TensorExprKernel>(
+  //         node->g(attr::Subgraph), custom_lowerings, sym_shapes);
+  //     cached_kernels[node] = kernel;
+  //   }
+  //   RECORD_FUNCTION(kernel->getKernelName(), std::vector<c10::IValue>());
+  //   kernel->run(stack);
+  //   return 0;
+  // };
+
+  // This also results in not using the cache at all.
+  // return [=](Stack& stack) {
+  //   static std::unordered_map<Graph*, std::shared_ptr<tensorexpr::TensorExprKernel>> cached_kernels;
+  //   auto subgraph = node->g(attr::Subgraph);
+  //   std::cout << "-- In createTensorExprOp: " << node << " : " << *node << std::endl;
+  //   std::cout << "  -- subgraph.get(): " << subgraph.get() << std::endl;
+  //   auto it = cached_kernels.find(subgraph.get());
+  //   std::shared_ptr<tensorexpr::TensorExprKernel> kernel;
+  //   if (it != cached_kernels.end()) {
+  //     kernel = it->second;
+  //     std::cout << "---- Reusing existing kernel" << std::endl;
+  //   } else {
+  //     std::cout << "---- Compiling a new kernel" << std::endl;
+  //     auto sym_shapes = node->is(attr::symbolic_shape_inputs);
+  //     std::unordered_map<c10::Symbol, tensorexpr::NNCLoweringFunction>
+  //         custom_lowerings;
+  //     kernel = std::make_shared<tensorexpr::TensorExprKernel>(
+  //         subgraph, custom_lowerings, sym_shapes);
+  //     cached_kernels[subgraph.get()] = kernel;
+  //   }
+  //   RECORD_FUNCTION(kernel->getKernelName(), std::vector<c10::IValue>());
+  //   kernel->run(stack);
+  //   return 0;
+  // };
+
+  return [=](Stack& stack) {
+    static std::unordered_map<std::string, std::shared_ptr<tensorexpr::TensorExprKernel>> cached_kernels;
```

Comment:
is `cached_kernels` accessed from multiple threads?

Reply:
Yes. In the SR workflow, that can potentially happen. I will have to update this to be thread-safe.